### PR TITLE
Or pattern types & parsing

### PR DESCRIPTION
**Pull request by reilem**

### CODE REVIEW

File: Frontend/HsParser.hs [hunk 1]
```diff
@@ -265,17 +265,25 @@ pPrimPat =  PsdWildPat <$  symbol "_"
 
 -- Parse a parsed pattern (lowest priority)
 pPatMain :: PsM PsdPat
-pPatMain = chainl1 pPrimPat (pure PsdAppPat)
+pPatMain = chainl1 (chainl1 pPrimPat (pure PsdAppPat)) (pPatOr <$ symbol "||")
+
+-- Create a parsed or pattern
+pPatOr :: PsdPat -> PsdPat -> PsdPat
+pPatOr p1 p2 = PsdOrPat p1 p2
```
Comment (gkaracha) [inline Frontend/HsParser.hs]: You can remove `pPatOr` and use `PsdOrPat` directly. General piece of advice: only define "smart" constructors if (a) you really want to hide the data constructors, or (b) if they do some actual work (e.g. creating an arrow type given a list of types). If the implementation of the "smart" function is identical to the data constructor and you are not hiding the constructor then you are just making the code longer ;-)

File: Frontend/HsTypes.hs [hunk 2]
```diff
@@ -143,19 +145,23 @@ instance (Symable a, PrettyPrint a) => PrettyPrint (HsAlt a) where
 instance (Symable a, PrettyPrint a) => PrettyPrint (HsPat a) where
   ppr (HsConPat dc xs)        = ppr dc <+> hsep (map pprPar xs)
   ppr (HsVarPat x    )        = ppr x
+  ppr (HsOrPat  p1 p2)        = pprPar p1 <+> text "||" <+> pprPar p2
   ppr HsWildPat               = text "_"
   needsParens (HsConPat _ xs) = length xs > 0
   needsParens (HsVarPat _   ) = False
+  needsParens (HsOrPat  _  _) = False
```
Comment (gkaracha) [inline Frontend/HsTypes.hs:152]: We might have to revise this. It's OK for now I think.

**Review by gkaracha — CHANGES_REQUESTED** [review]
Just inline `pPatOr` and feel free to merge!

File: Frontend/HsParser.hs (hunk 1 above)
Comment (reilem) [reply Frontend/HsParser.hs]: Oh yeah wow, I just realised how derpy this is 😄

**Pull request merged**